{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "BiocManager::install('chromVAR')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# install.packages(\"devtools\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Downloading GitHub repo xindong95/chromVAR@HEAD\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IRanges      (2.22.1   -> 2.22.2  ) [CRAN]\n",
      "S4Vectors    (0.26.0   -> 0.26.1  ) [CRAN]\n",
      "GenomeInfoDb (1.24.0   -> 1.24.2  ) [CRAN]\n",
      "Rcpp         (1.0.4.6  -> 1.0.5   ) [CRAN]\n",
      "cpp11        (NA       -> 0.1.0   ) [CRAN]\n",
      "dplyr        (1.0.0    -> 1.0.1   ) [CRAN]\n",
      "DelayedArray (0.14.0   -> 0.14.1  ) [CRAN]\n",
      "Summarize... (1.18.1   -> 1.18.2  ) [CRAN]\n",
      "XML          (3.99-0.3 -> 3.99-0.5) [CRAN]\n",
      "bit          (1.1-15.2 -> 4.0.3   ) [CRAN]\n",
      "bit64        (0.9-7.1  -> 4.0.2   ) [CRAN]\n",
      "Annotatio... (1.50.0   -> 1.50.3  ) [CRAN]\n",
      "data.table   (1.12.8   -> 1.13.0  ) [CRAN]\n",
      "tidyr        (1.1.0    -> 1.1.1   ) [CRAN]\n",
      "seqLogo      (1.54.1   -> 1.54.3  ) [CRAN]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Installing 15 packages: IRanges, S4Vectors, GenomeInfoDb, Rcpp, cpp11, dplyr, DelayedArray, SummarizedExperiment, XML, bit, bit64, AnnotationDbi, data.table, tidyr, seqLogo\n",
      "\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32m✔\u001b[39m  \u001b[38;5;247mchecking for file ‘/tmp/RtmpHDkpRc/remotes59b0374c0db7/xindong95-chromVAR-a888e57/DESCRIPTION’\u001b[39m\u001b[36m\u001b[36m (478ms)\u001b[36m\u001b[39m\n",
      "\u001b[38;5;247m─\u001b[39m\u001b[38;5;247m  \u001b[39m\u001b[38;5;247mpreparing ‘chromVAR’:\u001b[39m\u001b[36m\u001b[39m\n",
      "\u001b[32m✔\u001b[39m  \u001b[38;5;247mchecking DESCRIPTION meta-information\u001b[39m\u001b[36m\u001b[39m\n",
      "\u001b[38;5;247m─\u001b[39m\u001b[38;5;247m  \u001b[39m\u001b[38;5;247mcleaning src\u001b[39m\u001b[36m\u001b[39m\n",
      "\u001b[38;5;247m─\u001b[39m\u001b[38;5;247m  \u001b[39m\u001b[38;5;247mchecking for LF line-endings in source and make files and shell scripts\u001b[39m\u001b[36m\u001b[39m\n",
      "\u001b[38;5;247m─\u001b[39m\u001b[38;5;247m  \u001b[39m\u001b[38;5;247mchecking for empty or unneeded directories\u001b[39m\u001b[36m\u001b[39m\n",
      "\u001b[38;5;247m─\u001b[39m\u001b[38;5;247m  \u001b[39m\u001b[38;5;247mbuilding ‘chromVAR_1.5.0.tar.gz’\u001b[39m\u001b[36m\u001b[39m\n",
      "   \n",
      "\r"
     ]
    }
   ],
   "source": [
    "devtools::install_github(\"xindong95/chromVAR\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Registered S3 method overwritten by 'GGally':\n",
      "  method from   \n",
      "  +.gg   ggplot2\n",
      "\n",
      "Registered S3 method overwritten by 'cli':\n",
      "  method     from         \n",
      "  print.boxx spatstat.geom\n",
      "\n",
      "Warning message:\n",
      "\"package 'chromVAR' was built under R version 4.0.3\"\n",
      "\n",
      "\n",
      "Warning message:\n",
      "\"package 'Seurat' was built under R version 4.0.3\"\n",
      "Attaching SeuratObject\n",
      "\n",
      "Loading required package: BSgenome\n",
      "\n",
      "Warning message:\n",
      "\"package 'BSgenome' was built under R version 4.0.3\"\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Loading required package: parallel\n",
      "\n",
      "\n",
      "Attaching package: 'BiocGenerics'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:parallel':\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:stats':\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "\n",
      "Attaching package: 'S4Vectors'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:Matrix':\n",
      "\n",
      "    expand\n",
      "\n",
      "\n",
      "The following object is masked from 'package:base':\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Warning message:\n",
      "\"package 'GenomicRanges' was built under R version 4.0.3\"\n",
      "Loading required package: Biostrings\n",
      "\n",
      "Warning message:\n",
      "\"package 'Biostrings' was built under R version 4.0.3\"\n",
      "Loading required package: XVector\n",
      "\n",
      "Warning message:\n",
      "\"package 'XVector' was built under R version 4.0.3\"\n",
      "\n",
      "Attaching package: 'Biostrings'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:base':\n",
      "\n",
      "    strsplit\n",
      "\n",
      "\n",
      "Loading required package: rtracklayer\n",
      "\n",
      "Warning message:\n",
      "\"package 'rtracklayer' was built under R version 4.0.3\"\n",
      "Warning message:\n",
      "\"package 'motifmatchr' was built under R version 4.0.3\"\n",
      "Warning message:\n",
      "\"package 'SummarizedExperiment' was built under R version 4.0.3\"\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Warning message:\n",
      "\"package 'MatrixGenerics' was built under R version 4.0.3\"\n",
      "Loading required package: matrixStats\n",
      "\n",
      "Warning message:\n",
      "\"package 'matrixStats' was built under R version 4.0.3\"\n",
      "\n",
      "Attaching package: 'MatrixGenerics'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:matrixStats':\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Warning message:\n",
      "\"package 'Biobase' was built under R version 4.0.3\"\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'Biobase'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:MatrixGenerics':\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:matrixStats':\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'SummarizedExperiment'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:SeuratObject':\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "The following object is masked from 'package:Seurat':\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n"
     ]
    },
    {
     "ename": "ERROR",
     "evalue": "Error in library(rhdf5): there is no package called 'rhdf5'\n",
     "output_type": "error",
     "traceback": [
      "Error in library(rhdf5): there is no package called 'rhdf5'\nTraceback:\n",
      "1. library(rhdf5)"
     ]
    }
   ],
   "source": [
    "library(MAESTRO)\n",
    "library(chromVAR)\n",
    "library(Seurat)\n",
    "library(Matrix)\n",
    "library(BSgenome.Hsapiens.UCSC.hg38)\n",
    "library(BiocParallel)\n",
    "library(motifmatchr)\n",
    "library(SummarizedExperiment)\n",
    "library(rhdf5)\n",
    "library(JASPAR2016)\n",
    "register(MulticoreParam(8))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [],
   "source": [
    "chromVar <- function(peakfile, bamfile, celltype){\n",
    "    peaks <- readNarrowpeaks(peakfile,non_overlapping = FALSE)\n",
    "    fragment_counts <- getCounts(bamfile, \n",
    "                                 peaks, \n",
    "                                 paired =  TRUE, \n",
    "                                 by_rg = TRUE, \n",
    "                                 format = \"bam\", \n",
    "                                 colData = DataFrame(celltype = celltype))\n",
    "\n",
    "    motifs <- getJasparMotifs()\n",
    "    motif_ix <- matchMotifs(motifs, fragment_counts, \n",
    "                            genome = BSgenome.Hsapiens.UCSC.hg38)\n",
    "    return(motif_ix)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "peakfile <- \"analysis/GSE132065/GSM4499023/Result/Analysis/GSM4499023_all_peaks_regular.narrowPeak\"\n",
    "bamfile <- \"analysis/GSE132065/GSM4499023/Result/minimap2/GSM4499023.sortedByPos.rmdp.CBadded.unique.bam\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "peaks <- readNarrowpeaks(peakfile,non_overlapping = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Reading in file: analysis/GSE132065/GSM4499023/Result/minimap2/GSM4499023.sortedByPos.rmdp.CBadded.unique.bam\n",
      "\n"
     ]
    }
   ],
   "source": [
    "fragment_counts <- getCounts(bamfile, peaks, paired =  TRUE, by_rg = TRUE, format = \"bam\", colData = DataFrame(celltype = celltype))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Parsed with column specification:\n",
      "cols(\n",
      "  chr = \u001b[31mcol_character()\u001b[39m,\n",
      "  start = \u001b[32mcol_double()\u001b[39m,\n",
      "  end = \u001b[32mcol_double()\u001b[39m,\n",
      "  name = \u001b[31mcol_character()\u001b[39m,\n",
      "  score = \u001b[32mcol_double()\u001b[39m,\n",
      "  strand = \u001b[31mcol_character()\u001b[39m,\n",
      "  fc = \u001b[32mcol_double()\u001b[39m,\n",
      "  pval = \u001b[32mcol_double()\u001b[39m,\n",
      "  qval = \u001b[32mcol_double()\u001b[39m,\n",
      "  summit = \u001b[32mcol_double()\u001b[39m\n",
      ")\n",
      "\n"
     ]
    }
   ],
   "source": [
    "GSM4499023_motif <- chromVar(\"analysis/GSE132065/GSM4499023/Result/Analysis/GSM4499023_all_peaks_regular.narrowPeak\", \"analysis/GSE132065/GSM4499023/Result/minimap2/GSM4499023.sortedByPos.rmdp.CBadded.unique.bed\", \"CLL\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Parsed with column specification:\n",
      "cols(\n",
      "  chr = \u001b[31mcol_character()\u001b[39m,\n",
      "  start = \u001b[32mcol_double()\u001b[39m,\n",
      "  end = \u001b[32mcol_double()\u001b[39m,\n",
      "  name = \u001b[31mcol_character()\u001b[39m,\n",
      "  score = \u001b[32mcol_double()\u001b[39m,\n",
      "  strand = \u001b[31mcol_character()\u001b[39m,\n",
      "  fc = \u001b[32mcol_double()\u001b[39m,\n",
      "  pval = \u001b[32mcol_double()\u001b[39m,\n",
      "  qval = \u001b[32mcol_double()\u001b[39m,\n",
      "  summit = \u001b[32mcol_double()\u001b[39m\n",
      ")\n",
      "\n",
      "Parsed with column specification:\n",
      "cols(\n",
      "  chr = \u001b[31mcol_character()\u001b[39m,\n",
      "  start = \u001b[32mcol_double()\u001b[39m,\n",
      "  end = \u001b[32mcol_double()\u001b[39m,\n",
      "  name = \u001b[31mcol_character()\u001b[39m,\n",
      "  score = \u001b[32mcol_double()\u001b[39m,\n",
      "  strand = \u001b[31mcol_character()\u001b[39m,\n",
      "  fc = \u001b[32mcol_double()\u001b[39m,\n",
      "  pval = \u001b[32mcol_double()\u001b[39m,\n",
      "  qval = \u001b[32mcol_double()\u001b[39m,\n",
      "  summit = \u001b[32mcol_double()\u001b[39m\n",
      ")\n",
      "\n",
      "Parsed with column specification:\n",
      "cols(\n",
      "  chr = \u001b[31mcol_character()\u001b[39m,\n",
      "  start = \u001b[32mcol_double()\u001b[39m,\n",
      "  end = \u001b[32mcol_double()\u001b[39m,\n",
      "  name = \u001b[31mcol_character()\u001b[39m,\n",
      "  score = \u001b[32mcol_double()\u001b[39m,\n",
      "  strand = \u001b[31mcol_character()\u001b[39m,\n",
      "  fc = \u001b[32mcol_double()\u001b[39m,\n",
      "  pval = \u001b[32mcol_double()\u001b[39m,\n",
      "  qval = \u001b[32mcol_double()\u001b[39m,\n",
      "  summit = \u001b[32mcol_double()\u001b[39m\n",
      ")\n",
      "\n",
      "Parsed with column specification:\n",
      "cols(\n",
      "  chr = \u001b[31mcol_character()\u001b[39m,\n",
      "  start = \u001b[32mcol_double()\u001b[39m,\n",
      "  end = \u001b[32mcol_double()\u001b[39m,\n",
      "  name = \u001b[31mcol_character()\u001b[39m,\n",
      "  score = \u001b[32mcol_double()\u001b[39m,\n",
      "  strand = \u001b[31mcol_character()\u001b[39m,\n",
      "  fc = \u001b[32mcol_double()\u001b[39m,\n",
      "  pval = \u001b[32mcol_double()\u001b[39m,\n",
      "  qval = \u001b[32mcol_double()\u001b[39m,\n",
      "  summit = \u001b[32mcol_double()\u001b[39m\n",
      ")\n",
      "\n",
      "Parsed with column specification:\n",
      "cols(\n",
      "  chr = \u001b[31mcol_character()\u001b[39m,\n",
      "  start = \u001b[32mcol_double()\u001b[39m,\n",
      "  end = \u001b[32mcol_double()\u001b[39m,\n",
      "  name = \u001b[31mcol_character()\u001b[39m,\n",
      "  score = \u001b[32mcol_double()\u001b[39m,\n",
      "  strand = \u001b[31mcol_character()\u001b[39m,\n",
      "  fc = \u001b[32mcol_double()\u001b[39m,\n",
      "  pval = \u001b[32mcol_double()\u001b[39m,\n",
      "  qval = \u001b[32mcol_double()\u001b[39m,\n",
      "  summit = \u001b[32mcol_double()\u001b[39m\n",
      ")\n",
      "\n"
     ]
    }
   ],
   "source": [
    "GSM4499024_motif <- chromVar(\"analysis/GSE132065/GSM4499024/Result/Analysis/GSM4499024_all_peaks_regular.narrowPeak\", \"analysis/GSE132065/GSM4499024/Result/minimap2/GSM4499024.sortedByPos.rmdp.CBadded.unique.bed\", \"CLL\")\n",
    "\n",
    "GSM4499025_motif <- chromVar(\"analysis/GSE132065/GSM4499025/Result/Analysis/GSM4499025_all_peaks_regular.narrowPeak\", \"analysis/GSE132065/GSM4499025/Result/minimap2/GSM4499025.sortedByPos.rmdp.CBadded.unique.bed\", \"CLL\")\n",
    "\n",
    "GSM4499026_motif <- chromVar(\"analysis/GSE132065/GSM4499026/Result/Analysis/GSM4499026_all_peaks_regular.narrowPeak\", \"analysis/GSE132065/GSM4499026/Result/minimap2/GSM4499026.sortedByPos.rmdp.CBadded.unique.bed\", \"PBMC\")\n",
    "\n",
    "GSM4499027_motif <- chromVar(\"analysis/GSE132065/GSM4499027/Result/Analysis/GSM4499027_all_peaks_regular.narrowPeak\", \"analysis/GSE132065/GSM4499027/Result/minimap2/GSM4499027.sortedByPos.rmdp.CBadded.unique.bed\", \"PBMC\")\n",
    "\n",
    "GSM4499028_motif <- chromVar(\"analysis/GSE132065/GSM4499028/Result/Analysis/GSM4499028_all_peaks_regular.narrowPeak\", \"analysis/GSE132065/GSM4499028/Result/minimap2/GSM4499028.sortedByPos.rmdp.CBadded.unique.bed\", \"PBMC\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A matrix: 386 × 1 of type chr</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>name</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>MA0025.1_NFIL3</th><td>NFIL3      </td></tr>\n",
       "\t<tr><th scope=row>MA0030.1_FOXF2</th><td>FOXF2      </td></tr>\n",
       "\t<tr><th scope=row>MA0031.1_FOXD1</th><td>FOXD1      </td></tr>\n",
       "\t<tr><th scope=row>MA0051.1_IRF2</th><td>IRF2       </td></tr>\n",
       "\t<tr><th scope=row>MA0056.1_MZF1</th><td>MZF1       </td></tr>\n",
       "\t<tr><th scope=row>MA0057.1_MZF1(var.2)</th><td>MZF1(var.2)</td></tr>\n",
       "\t<tr><th scope=row>MA0059.1_MAX::MYC</th><td>MAX::MYC   </td></tr>\n",
       "\t<tr><th scope=row>MA0066.1_PPARG</th><td>PPARG      </td></tr>\n",
       "\t<tr><th scope=row>MA0069.1_Pax6</th><td>Pax6       </td></tr>\n",
       "\t<tr><th scope=row>MA0070.1_PBX1</th><td>PBX1       </td></tr>\n",
       "\t<tr><th scope=row>MA0071.1_RORA</th><td>RORA       </td></tr>\n",
       "\t<tr><th scope=row>MA0072.1_RORA(var.2)</th><td>RORA(var.2)</td></tr>\n",
       "\t<tr><th scope=row>MA0073.1_RREB1</th><td>RREB1      </td></tr>\n",
       "\t<tr><th scope=row>MA0074.1_RXRA::VDR</th><td>RXRA::VDR  </td></tr>\n",
       "\t<tr><th scope=row>MA0077.1_SOX9</th><td>SOX9       </td></tr>\n",
       "\t<tr><th scope=row>MA0081.1_SPIB</th><td>SPIB       </td></tr>\n",
       "\t<tr><th scope=row>MA0084.1_SRY</th><td>SRY        </td></tr>\n",
       "\t<tr><th scope=row>MA0091.1_TAL1::TCF3</th><td>TAL1::TCF3 </td></tr>\n",
       "\t<tr><th scope=row>MA0101.1_REL</th><td>REL        </td></tr>\n",
       "\t<tr><th scope=row>MA0107.1_RELA</th><td>RELA       </td></tr>\n",
       "\t<tr><th scope=row>MA0115.1_NR1H2::RXRA</th><td>NR1H2::RXRA</td></tr>\n",
       "\t<tr><th scope=row>MA0119.1_NFIC::TLX1</th><td>NFIC::TLX1 </td></tr>\n",
       "\t<tr><th scope=row>MA0130.1_ZNF354C</th><td>ZNF354C    </td></tr>\n",
       "\t<tr><th scope=row>MA0139.1_CTCF</th><td>CTCF       </td></tr>\n",
       "\t<tr><th scope=row>MA0149.1_EWSR1-FLI1</th><td>EWSR1-FLI1 </td></tr>\n",
       "\t<tr><th scope=row>MA0138.2_REST</th><td>REST       </td></tr>\n",
       "\t<tr><th scope=row>MA0152.1_NFATC2</th><td>NFATC2     </td></tr>\n",
       "\t<tr><th scope=row>MA0155.1_INSM1</th><td>INSM1      </td></tr>\n",
       "\t<tr><th scope=row>MA0158.1_HOXA5</th><td>HOXA5      </td></tr>\n",
       "\t<tr><th scope=row>MA0159.1_RARA::RXRA</th><td>RARA::RXRA </td></tr>\n",
       "\t<tr><th scope=row>⋮</th><td>⋮</td></tr>\n",
       "\t<tr><th scope=row>MA0871.1_TFEC</th><td>TFEC         </td></tr>\n",
       "\t<tr><th scope=row>MA0872.1_TFAP2A(var.3)</th><td>TFAP2A(var.3)</td></tr>\n",
       "\t<tr><th scope=row>MA0028.2_ELK1</th><td>ELK1         </td></tr>\n",
       "\t<tr><th scope=row>MA0873.1_HOXD12</th><td>HOXD12       </td></tr>\n",
       "\t<tr><th scope=row>MA0875.1_BARX1</th><td>BARX1        </td></tr>\n",
       "\t<tr><th scope=row>MA0876.1_BSX</th><td>BSX          </td></tr>\n",
       "\t<tr><th scope=row>MA0878.1_CDX1</th><td>CDX1         </td></tr>\n",
       "\t<tr><th scope=row>MA0882.1_DLX6</th><td>DLX6         </td></tr>\n",
       "\t<tr><th scope=row>MA0884.1_DUXA</th><td>DUXA         </td></tr>\n",
       "\t<tr><th scope=row>MA0886.1_EMX2</th><td>EMX2         </td></tr>\n",
       "\t<tr><th scope=row>MA0887.1_EVX1</th><td>EVX1         </td></tr>\n",
       "\t<tr><th scope=row>MA0888.1_EVX2</th><td>EVX2         </td></tr>\n",
       "\t<tr><th scope=row>MA0889.1_GBX1</th><td>GBX1         </td></tr>\n",
       "\t<tr><th scope=row>MA0890.1_GBX2</th><td>GBX2         </td></tr>\n",
       "\t<tr><th scope=row>MA0891.1_GSC2</th><td>GSC2         </td></tr>\n",
       "\t<tr><th scope=row>MA0892.1_GSX1</th><td>GSX1         </td></tr>\n",
       "\t<tr><th scope=row>MA0893.1_GSX2</th><td>GSX2         </td></tr>\n",
       "\t<tr><th scope=row>MA0894.1_HESX1</th><td>HESX1        </td></tr>\n",
       "\t<tr><th scope=row>MA0895.1_HMBOX1</th><td>HMBOX1       </td></tr>\n",
       "\t<tr><th scope=row>MA0899.1_HOXA10</th><td>HOXA10       </td></tr>\n",
       "\t<tr><th scope=row>MA0900.1_HOXA2</th><td>HOXA2        </td></tr>\n",
       "\t<tr><th scope=row>MA0901.1_HOXB13</th><td>HOXB13       </td></tr>\n",
       "\t<tr><th scope=row>MA0902.1_HOXB2</th><td>HOXB2        </td></tr>\n",
       "\t<tr><th scope=row>MA0903.1_HOXB3</th><td>HOXB3        </td></tr>\n",
       "\t<tr><th scope=row>MA0905.1_HOXC10</th><td>HOXC10       </td></tr>\n",
       "\t<tr><th scope=row>MA0906.1_HOXC12</th><td>HOXC12       </td></tr>\n",
       "\t<tr><th scope=row>MA0907.1_HOXC13</th><td>HOXC13       </td></tr>\n",
       "\t<tr><th scope=row>MA0908.1_HOXD11</th><td>HOXD11       </td></tr>\n",
       "\t<tr><th scope=row>MA0909.1_HOXD13</th><td>HOXD13       </td></tr>\n",
       "\t<tr><th scope=row>MA0914.1_ISL2</th><td>ISL2         </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 386 × 1 of type chr\n",
       "\\begin{tabular}{r|l}\n",
       "  & name\\\\\n",
       "\\hline\n",
       "\tMA0025.1\\_NFIL3 & NFIL3      \\\\\n",
       "\tMA0030.1\\_FOXF2 & FOXF2      \\\\\n",
       "\tMA0031.1\\_FOXD1 & FOXD1      \\\\\n",
       "\tMA0051.1\\_IRF2 & IRF2       \\\\\n",
       "\tMA0056.1\\_MZF1 & MZF1       \\\\\n",
       "\tMA0057.1\\_MZF1(var.2) & MZF1(var.2)\\\\\n",
       "\tMA0059.1\\_MAX::MYC & MAX::MYC   \\\\\n",
       "\tMA0066.1\\_PPARG & PPARG      \\\\\n",
       "\tMA0069.1\\_Pax6 & Pax6       \\\\\n",
       "\tMA0070.1\\_PBX1 & PBX1       \\\\\n",
       "\tMA0071.1\\_RORA & RORA       \\\\\n",
       "\tMA0072.1\\_RORA(var.2) & RORA(var.2)\\\\\n",
       "\tMA0073.1\\_RREB1 & RREB1      \\\\\n",
       "\tMA0074.1\\_RXRA::VDR & RXRA::VDR  \\\\\n",
       "\tMA0077.1\\_SOX9 & SOX9       \\\\\n",
       "\tMA0081.1\\_SPIB & SPIB       \\\\\n",
       "\tMA0084.1\\_SRY & SRY        \\\\\n",
       "\tMA0091.1\\_TAL1::TCF3 & TAL1::TCF3 \\\\\n",
       "\tMA0101.1\\_REL & REL        \\\\\n",
       "\tMA0107.1\\_RELA & RELA       \\\\\n",
       "\tMA0115.1\\_NR1H2::RXRA & NR1H2::RXRA\\\\\n",
       "\tMA0119.1\\_NFIC::TLX1 & NFIC::TLX1 \\\\\n",
       "\tMA0130.1\\_ZNF354C & ZNF354C    \\\\\n",
       "\tMA0139.1\\_CTCF & CTCF       \\\\\n",
       "\tMA0149.1\\_EWSR1-FLI1 & EWSR1-FLI1 \\\\\n",
       "\tMA0138.2\\_REST & REST       \\\\\n",
       "\tMA0152.1\\_NFATC2 & NFATC2     \\\\\n",
       "\tMA0155.1\\_INSM1 & INSM1      \\\\\n",
       "\tMA0158.1\\_HOXA5 & HOXA5      \\\\\n",
       "\tMA0159.1\\_RARA::RXRA & RARA::RXRA \\\\\n",
       "\t⋮ & ⋮\\\\\n",
       "\tMA0871.1\\_TFEC & TFEC         \\\\\n",
       "\tMA0872.1\\_TFAP2A(var.3) & TFAP2A(var.3)\\\\\n",
       "\tMA0028.2\\_ELK1 & ELK1         \\\\\n",
       "\tMA0873.1\\_HOXD12 & HOXD12       \\\\\n",
       "\tMA0875.1\\_BARX1 & BARX1        \\\\\n",
       "\tMA0876.1\\_BSX & BSX          \\\\\n",
       "\tMA0878.1\\_CDX1 & CDX1         \\\\\n",
       "\tMA0882.1\\_DLX6 & DLX6         \\\\\n",
       "\tMA0884.1\\_DUXA & DUXA         \\\\\n",
       "\tMA0886.1\\_EMX2 & EMX2         \\\\\n",
       "\tMA0887.1\\_EVX1 & EVX1         \\\\\n",
       "\tMA0888.1\\_EVX2 & EVX2         \\\\\n",
       "\tMA0889.1\\_GBX1 & GBX1         \\\\\n",
       "\tMA0890.1\\_GBX2 & GBX2         \\\\\n",
       "\tMA0891.1\\_GSC2 & GSC2         \\\\\n",
       "\tMA0892.1\\_GSX1 & GSX1         \\\\\n",
       "\tMA0893.1\\_GSX2 & GSX2         \\\\\n",
       "\tMA0894.1\\_HESX1 & HESX1        \\\\\n",
       "\tMA0895.1\\_HMBOX1 & HMBOX1       \\\\\n",
       "\tMA0899.1\\_HOXA10 & HOXA10       \\\\\n",
       "\tMA0900.1\\_HOXA2 & HOXA2        \\\\\n",
       "\tMA0901.1\\_HOXB13 & HOXB13       \\\\\n",
       "\tMA0902.1\\_HOXB2 & HOXB2        \\\\\n",
       "\tMA0903.1\\_HOXB3 & HOXB3        \\\\\n",
       "\tMA0905.1\\_HOXC10 & HOXC10       \\\\\n",
       "\tMA0906.1\\_HOXC12 & HOXC12       \\\\\n",
       "\tMA0907.1\\_HOXC13 & HOXC13       \\\\\n",
       "\tMA0908.1\\_HOXD11 & HOXD11       \\\\\n",
       "\tMA0909.1\\_HOXD13 & HOXD13       \\\\\n",
       "\tMA0914.1\\_ISL2 & ISL2         \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 386 × 1 of type chr\n",
       "\n",
       "| <!--/--> | name |\n",
       "|---|---|\n",
       "| MA0025.1_NFIL3 | NFIL3       |\n",
       "| MA0030.1_FOXF2 | FOXF2       |\n",
       "| MA0031.1_FOXD1 | FOXD1       |\n",
       "| MA0051.1_IRF2 | IRF2        |\n",
       "| MA0056.1_MZF1 | MZF1        |\n",
       "| MA0057.1_MZF1(var.2) | MZF1(var.2) |\n",
       "| MA0059.1_MAX::MYC | MAX::MYC    |\n",
       "| MA0066.1_PPARG | PPARG       |\n",
       "| MA0069.1_Pax6 | Pax6        |\n",
       "| MA0070.1_PBX1 | PBX1        |\n",
       "| MA0071.1_RORA | RORA        |\n",
       "| MA0072.1_RORA(var.2) | RORA(var.2) |\n",
       "| MA0073.1_RREB1 | RREB1       |\n",
       "| MA0074.1_RXRA::VDR | RXRA::VDR   |\n",
       "| MA0077.1_SOX9 | SOX9        |\n",
       "| MA0081.1_SPIB | SPIB        |\n",
       "| MA0084.1_SRY | SRY         |\n",
       "| MA0091.1_TAL1::TCF3 | TAL1::TCF3  |\n",
       "| MA0101.1_REL | REL         |\n",
       "| MA0107.1_RELA | RELA        |\n",
       "| MA0115.1_NR1H2::RXRA | NR1H2::RXRA |\n",
       "| MA0119.1_NFIC::TLX1 | NFIC::TLX1  |\n",
       "| MA0130.1_ZNF354C | ZNF354C     |\n",
       "| MA0139.1_CTCF | CTCF        |\n",
       "| MA0149.1_EWSR1-FLI1 | EWSR1-FLI1  |\n",
       "| MA0138.2_REST | REST        |\n",
       "| MA0152.1_NFATC2 | NFATC2      |\n",
       "| MA0155.1_INSM1 | INSM1       |\n",
       "| MA0158.1_HOXA5 | HOXA5       |\n",
       "| MA0159.1_RARA::RXRA | RARA::RXRA  |\n",
       "| ⋮ | ⋮ |\n",
       "| MA0871.1_TFEC | TFEC          |\n",
       "| MA0872.1_TFAP2A(var.3) | TFAP2A(var.3) |\n",
       "| MA0028.2_ELK1 | ELK1          |\n",
       "| MA0873.1_HOXD12 | HOXD12        |\n",
       "| MA0875.1_BARX1 | BARX1         |\n",
       "| MA0876.1_BSX | BSX           |\n",
       "| MA0878.1_CDX1 | CDX1          |\n",
       "| MA0882.1_DLX6 | DLX6          |\n",
       "| MA0884.1_DUXA | DUXA          |\n",
       "| MA0886.1_EMX2 | EMX2          |\n",
       "| MA0887.1_EVX1 | EVX1          |\n",
       "| MA0888.1_EVX2 | EVX2          |\n",
       "| MA0889.1_GBX1 | GBX1          |\n",
       "| MA0890.1_GBX2 | GBX2          |\n",
       "| MA0891.1_GSC2 | GSC2          |\n",
       "| MA0892.1_GSX1 | GSX1          |\n",
       "| MA0893.1_GSX2 | GSX2          |\n",
       "| MA0894.1_HESX1 | HESX1         |\n",
       "| MA0895.1_HMBOX1 | HMBOX1        |\n",
       "| MA0899.1_HOXA10 | HOXA10        |\n",
       "| MA0900.1_HOXA2 | HOXA2         |\n",
       "| MA0901.1_HOXB13 | HOXB13        |\n",
       "| MA0902.1_HOXB2 | HOXB2         |\n",
       "| MA0903.1_HOXB3 | HOXB3         |\n",
       "| MA0905.1_HOXC10 | HOXC10        |\n",
       "| MA0906.1_HOXC12 | HOXC12        |\n",
       "| MA0907.1_HOXC13 | HOXC13        |\n",
       "| MA0908.1_HOXD11 | HOXD11        |\n",
       "| MA0909.1_HOXD13 | HOXD13        |\n",
       "| MA0914.1_ISL2 | ISL2          |\n",
       "\n"
      ],
      "text/plain": [
       "                       name         \n",
       "MA0025.1_NFIL3         NFIL3        \n",
       "MA0030.1_FOXF2         FOXF2        \n",
       "MA0031.1_FOXD1         FOXD1        \n",
       "MA0051.1_IRF2          IRF2         \n",
       "MA0056.1_MZF1          MZF1         \n",
       "MA0057.1_MZF1(var.2)   MZF1(var.2)  \n",
       "MA0059.1_MAX::MYC      MAX::MYC     \n",
       "MA0066.1_PPARG         PPARG        \n",
       "MA0069.1_Pax6          Pax6         \n",
       "MA0070.1_PBX1          PBX1         \n",
       "MA0071.1_RORA          RORA         \n",
       "MA0072.1_RORA(var.2)   RORA(var.2)  \n",
       "MA0073.1_RREB1         RREB1        \n",
       "MA0074.1_RXRA::VDR     RXRA::VDR    \n",
       "MA0077.1_SOX9          SOX9         \n",
       "MA0081.1_SPIB          SPIB         \n",
       "MA0084.1_SRY           SRY          \n",
       "MA0091.1_TAL1::TCF3    TAL1::TCF3   \n",
       "MA0101.1_REL           REL          \n",
       "MA0107.1_RELA          RELA         \n",
       "MA0115.1_NR1H2::RXRA   NR1H2::RXRA  \n",
       "MA0119.1_NFIC::TLX1    NFIC::TLX1   \n",
       "MA0130.1_ZNF354C       ZNF354C      \n",
       "MA0139.1_CTCF          CTCF         \n",
       "MA0149.1_EWSR1-FLI1    EWSR1-FLI1   \n",
       "MA0138.2_REST          REST         \n",
       "MA0152.1_NFATC2        NFATC2       \n",
       "MA0155.1_INSM1         INSM1        \n",
       "MA0158.1_HOXA5         HOXA5        \n",
       "MA0159.1_RARA::RXRA    RARA::RXRA   \n",
       "⋮                      ⋮            \n",
       "MA0871.1_TFEC          TFEC         \n",
       "MA0872.1_TFAP2A(var.3) TFAP2A(var.3)\n",
       "MA0028.2_ELK1          ELK1         \n",
       "MA0873.1_HOXD12        HOXD12       \n",
       "MA0875.1_BARX1         BARX1        \n",
       "MA0876.1_BSX           BSX          \n",
       "MA0878.1_CDX1          CDX1         \n",
       "MA0882.1_DLX6          DLX6         \n",
       "MA0884.1_DUXA          DUXA         \n",
       "MA0886.1_EMX2          EMX2         \n",
       "MA0887.1_EVX1          EVX1         \n",
       "MA0888.1_EVX2          EVX2         \n",
       "MA0889.1_GBX1          GBX1         \n",
       "MA0890.1_GBX2          GBX2         \n",
       "MA0891.1_GSC2          GSC2         \n",
       "MA0892.1_GSX1          GSX1         \n",
       "MA0893.1_GSX2          GSX2         \n",
       "MA0894.1_HESX1         HESX1        \n",
       "MA0895.1_HMBOX1        HMBOX1       \n",
       "MA0899.1_HOXA10        HOXA10       \n",
       "MA0900.1_HOXA2         HOXA2        \n",
       "MA0901.1_HOXB13        HOXB13       \n",
       "MA0902.1_HOXB2         HOXB2        \n",
       "MA0903.1_HOXB3         HOXB3        \n",
       "MA0905.1_HOXC10        HOXC10       \n",
       "MA0906.1_HOXC12        HOXC12       \n",
       "MA0907.1_HOXC13        HOXC13       \n",
       "MA0908.1_HOXD11        HOXD11       \n",
       "MA0909.1_HOXD13        HOXD13       \n",
       "MA0914.1_ISL2          ISL2         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A matrix: 10 × 1 of type chr</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>name</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>MA0025.1_NFIL3</th><td>NFIL3      </td></tr>\n",
       "\t<tr><th scope=row>MA0030.1_FOXF2</th><td>FOXF2      </td></tr>\n",
       "\t<tr><th scope=row>MA0031.1_FOXD1</th><td>FOXD1      </td></tr>\n",
       "\t<tr><th scope=row>MA0051.1_IRF2</th><td>IRF2       </td></tr>\n",
       "\t<tr><th scope=row>MA0056.1_MZF1</th><td>MZF1       </td></tr>\n",
       "\t<tr><th scope=row>MA0057.1_MZF1(var.2)</th><td>MZF1(var.2)</td></tr>\n",
       "\t<tr><th scope=row>MA0059.1_MAX::MYC</th><td>MAX::MYC   </td></tr>\n",
       "\t<tr><th scope=row>MA0066.1_PPARG</th><td>PPARG      </td></tr>\n",
       "\t<tr><th scope=row>MA0069.1_Pax6</th><td>Pax6       </td></tr>\n",
       "\t<tr><th scope=row>MA0070.1_PBX1</th><td>PBX1       </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 10 × 1 of type chr\n",
       "\\begin{tabular}{r|l}\n",
       "  & name\\\\\n",
       "\\hline\n",
       "\tMA0025.1\\_NFIL3 & NFIL3      \\\\\n",
       "\tMA0030.1\\_FOXF2 & FOXF2      \\\\\n",
       "\tMA0031.1\\_FOXD1 & FOXD1      \\\\\n",
       "\tMA0051.1\\_IRF2 & IRF2       \\\\\n",
       "\tMA0056.1\\_MZF1 & MZF1       \\\\\n",
       "\tMA0057.1\\_MZF1(var.2) & MZF1(var.2)\\\\\n",
       "\tMA0059.1\\_MAX::MYC & MAX::MYC   \\\\\n",
       "\tMA0066.1\\_PPARG & PPARG      \\\\\n",
       "\tMA0069.1\\_Pax6 & Pax6       \\\\\n",
       "\tMA0070.1\\_PBX1 & PBX1       \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 10 × 1 of type chr\n",
       "\n",
       "| <!--/--> | name |\n",
       "|---|---|\n",
       "| MA0025.1_NFIL3 | NFIL3       |\n",
       "| MA0030.1_FOXF2 | FOXF2       |\n",
       "| MA0031.1_FOXD1 | FOXD1       |\n",
       "| MA0051.1_IRF2 | IRF2        |\n",
       "| MA0056.1_MZF1 | MZF1        |\n",
       "| MA0057.1_MZF1(var.2) | MZF1(var.2) |\n",
       "| MA0059.1_MAX::MYC | MAX::MYC    |\n",
       "| MA0066.1_PPARG | PPARG       |\n",
       "| MA0069.1_Pax6 | Pax6        |\n",
       "| MA0070.1_PBX1 | PBX1        |\n",
       "\n"
      ],
      "text/plain": [
       "                     name       \n",
       "MA0025.1_NFIL3       NFIL3      \n",
       "MA0030.1_FOXF2       FOXF2      \n",
       "MA0031.1_FOXD1       FOXD1      \n",
       "MA0051.1_IRF2        IRF2       \n",
       "MA0056.1_MZF1        MZF1       \n",
       "MA0057.1_MZF1(var.2) MZF1(var.2)\n",
       "MA0059.1_MAX::MYC    MAX::MYC   \n",
       "MA0066.1_PPARG       PPARG      \n",
       "MA0069.1_Pax6        Pax6       \n",
       "MA0070.1_PBX1        PBX1       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(as.matrix(colData(GSM4499026_motif)),10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DataFrame with 73403 rows and 3 columns\n",
       "          score      qval                      name\n",
       "      <numeric> <numeric>               <character>\n",
       "1          2396 239.64490     GSM4499026_all_peak_1\n",
       "2          1207 120.71465     GSM4499026_all_peak_2\n",
       "3            27   2.74408     GSM4499026_all_peak_3\n",
       "4            39   3.92265     GSM4499026_all_peak_4\n",
       "5           487  48.72235     GSM4499026_all_peak_5\n",
       "...         ...       ...                       ...\n",
       "73399        18   1.89537 GSM4499026_all_peak_74210\n",
       "73400        31   3.11868 GSM4499026_all_peak_74211\n",
       "73401        28   2.85361 GSM4499026_all_peak_74212\n",
       "73402        29   2.99665 GSM4499026_all_peak_74213\n",
       "73403        49   4.97791 GSM4499026_all_peak_74214"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "rowData(GSM4499026_motif)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "List of length 1\n",
       "names(1): motifMatches"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "assays(GSM4499026_motif)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Reading in file: /mnt/Storage/home/dongxin/Applications/miniconda3/envs/MAESTRO/lib/R/library/chromVAR/extdata/test_RG.bam\n",
      "\n"
     ]
    }
   ],
   "source": [
    "bamfile <- system.file(\"extdata/test_RG.bam\", package = \"chromVAR\")\n",
    "fragment_counts <- getCounts(bamfile, \n",
    "                             peaks, \n",
    "                             paired =  TRUE, \n",
    "                             by_rg = TRUE, \n",
    "                             format = \"bam\", \n",
    "                             colData = DataFrame(celltype = \"GM\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: RangedSummarizedExperiment \n",
       "dim: 78280 17 \n",
       "metadata(0):\n",
       "assays(1): counts\n",
       "rownames: NULL\n",
       "rowData names(3): score qval name\n",
       "colnames(17): singles-GM12878rep2-140904-14\n",
       "  singles-GM12878rep2-140904-16 ... singles-GM12878rep2-140904-93\n",
       "  singles-GM12878rep2-140904-94\n",
       "colData names(2): celltype depth"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fragment_counts"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R 4.0.2 MAESTRO",
   "language": "R",
   "name": "ir402maestro"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
